tests/unit_tests/scheduler/test_openpbs_driver.py

```diff
@@ -106,3 +112,112 @@ async def test_job_name_with_prefix():
+):
+    os.chdir(tmp_path)
+    bin_path = tmp_path / "bin"
+    bin_path.mkdir()
+    monkeypatch.setenv("PATH", f"{bin_path}:{os.environ['PATH']}")
+    qsub_path = bin_path / "qsub"
+    qsub_path.write_text(f"#!/bin/sh\necho {error_msg} >&2\nexit {exit_code}")
+    qsub_path.chmod(qsub_path.stat().st_mode | stat.S_IEXEC)
+    driver = OpenPBSDriver()
+    driver._num_retries = 2
+    driver._retry_interval = 0.2
+    match_str = (
+        f"failed after 2 retries with error {error_msg}"
+        if exit_code != 199
+        else "failed with exit code 199 and error message: Not recognized"
+    )
+    with pytest.raises(RuntimeError, match=match_str):
+        await driver.submit(0, "sleep 10")
+
+
+@pytest.mark.parametrize(
+    ("exit_code, error_msg"),
+    [
+        (QSUB_INVALID_CREDENTIAL, "Invalid credential"),
+        (QSUB_PREMATURE_END_OF_MESSAGE, "Premature end of message"),
+        (QSUB_CONNECTION_REFUSED, "Connection refused"),
+    ],
+)
+async def test_that_qsub_will_retry_and_succeed(
+    monkeypatch, tmp_path, exit_code, error_msg
+):
+    os.chdir(tmp_path)
+    bin_path = tmp_path / "bin"
+    bin_path.mkdir()
+    monkeypatch.setenv("PATH", f"{bin_path}:{os.environ['PATH']}")
+    qsub_path = bin_path / "qsub"
+    qsub_path.write_text(
+        "#!/bin/sh"
+        + dedent(
+            f"""
+            TRY_FILE="{bin_path}/script_try"
+            if [ -f "$TRY_FILE" ]; then
+                echo "SUCCESS"
+                exit 0
+            else
+                echo "TRIED" > $TRY_FILE
+                echo "{error_msg}" >&2
+                exit {exit_code}
+            fi
+            """
+        )
+    )
+    qsub_path.chmod(qsub_path.stat().st_mode | stat.S_IEXEC)
+    driver = OpenPBSDriver()
+    driver._num_retries = 2
+    driver._retry_interval = 0.2
+    await driver.submit(0, "sleep 10")
+
+
+@pytest.mark.parametrize(
+    ("exit_code, error_msg"),
+    [
+        (QDEL_JOB_HAS_FINISHED, "Job has finished"),
+        (QDEL_REQUEST_INVALID, "Request invalid for state of job"),
+    ],
+)
+async def test_that_qdel_will_retry_and_succeed(
+    monkeypatch, tmp_path, exit_code, error_msg
+):
+    os.chdir(tmp_path)
+    bin_path = tmp_path / "bin"
+    bin_path.mkdir()
+    monkeypatch.setenv("PATH", f"{bin_path}:{os.environ['PATH']}")
+    qdel_path = bin_path / "qdel"
+    qdel_path.write_text(
+        "#!/bin/sh"
+        + dedent(
+            f"""
+            TRY_FILE="{bin_path}/script_try"
+            if [ -f "$TRY_FILE" ]; then
+                echo "111"
+                exit 0
+            else
+                echo "TRIED" > $TRY_FILE
+                echo "{error_msg}" >&2
+                exit {exit_code}
+            fi
+            """
+        )
+    )
+    qdel_path.chmod(qdel_path.stat().st_mode | stat.S_IEXEC)
+    driver = OpenPBSDriver()
+    driver._num_retries = 2
+    driver._retry_interval = 0.2
+    driver._iens2jobid[0] = 111
+    await driver.kill(0)
```

Comment:
Is this testing anything at all? You can implement a very buggy `kill()` and this test will still pass.

Reply:
Yes, this will not raise an error. Not sure what to assert as this function returns `None`.

Reply:
If a function can't be tested, the problem is in the function..

Reply:
Updated!